src/utils/event-decorators.js

```diff
@@ -0,0 +1,163 @@
+ *
+ * @param p1 - If this is a function it will be decorated with default options and p2 is ignored. Otherwise this can be
+ *   an object which fine tunes the binding.
+ * @param p2 - If p1 is an object then this will be the function to bind to.
+ *
+ * @returns {function} Decorated function which wraps the input function.
+ */
+function bindEvent (p1, p2) {
+  if (typeof p1 === 'function') {
+    return decorate(p1, BindToEventDecorator());
+  } else if (typeof p1 === 'object' && typeof p2 === 'function') {
+    return decorate(p2, BindToEventDecorator(p1.event, p1.target, p1.listenIn, p1.removeIn));
+  } else {
+    throw new Error('bindEvent must take: (function), or a ([object], function)');
+  }
+}
+
+// Implements the automatic binding and unbinding of the chosen function. Wraps its listenIn and removeIn
+// functions to add and remove the event listener at the correct times.
+function BindToEventDecorator (_event, _target, _listenIn, _removeIn) {
+  return function (propertyName, func) {
+    var scope = this;
+    var event = _event || propertyName;
+    var target = !_target ? this.el : document.querySelector(_target);
+    if (!target) {
+      console.warn("Couldn't subscribe " + this.name + '.' + propertyName + ' to ' + event + ' on ' + _target +
+        ' because querySelector returned undefined.');
+      return;
+    }
+    var listenIn = _listenIn || 'init';
+    var removeIn = _removeIn || 'remove';
+
+    var listenFunc = this[listenIn];
+    var removeFunc = this[removeIn];
+    var boundFunc = func.bind(this);
+
+    this[listenIn] = function () {
+      if (listenFunc !== undefined) {
+        listenFunc.apply(scope, arguments);
+      }
+      target.addEventListener(event, boundFunc);
+    };
+
+    this[removeIn] = function () {
+      if (removeFunc !== undefined) {
+        removeFunc.apply(scope, arguments);
+      }
+      target.removeEventListener(event, boundFunc);
+    };
+
+    return func;
+  };
+}
+
+/*
+ * Decorates a function with a functor that is executed in the context of each instantiated Component which owns it.
+ *
+ * @param func - Function to decorate.
+ * @param decoratorFunc - Functor which decorates func. It's definition should appear like this:
+ *    function decoratorFunctor(funcPropertyName) {
+ *      var func = this[funcPropertyName]; // 'this' is assigned to the instantiated Component.
+ *      // Decorator should return a function
+ *      return function() {
+ *        console.log(this.el.id + " has a decorated function!");
+ *        return func.apply(this, arguments);
+ *      }
+ *    }
+ *
+ * @returns A pending decorated functor which will be executed when a component that owns it is instantiated.
+ */
+function decorate (func, decoratorFunc) {
+  return decorations.add(func, decoratorFunc);
+}
+
+/*
+ * Executes all decorated functions on the given component. You shouldn't need to call
+ * this yourself.
+ */
+function executeDecorators (component) {
```

Comment:
If this were called on every component, we'd want to optimize it. No closures (`forEach`), no array allocations.